addons/web/static/src/core/utils/template_inheritance.js

```diff
@@ -0,0 +1,196 @@
+/** @odoo-module **/
+
+import { _t } from "@web/core/l10n/translation";
+
+const parser = new DOMParser();
+const serializer = new XMLSerializer();
+
+// function rstrip(string) {
+//     return string.replace(/\s+$/g, "");
+// }
+
+// function getTextNode(nodes) {
+//     for (const child of nodes) {
+//         if (child.nodeType === 3) {
+//             return child;
+//         }
+//         if (child.nodeType === 1) {
+//             break;
+//         }
+//     }
+//     return null;
+// }
+
+function addBefore(target, operation) {
+    // manage text as expected in template_inheritance.py
+    const childNodes = [];
+    for (const childNode of operation.childNodes) {
+        if (childNode.tagName === "xpath" && childNode.getAttribute("position") === "move") {
+            const node = getNode(target.ownerDocument, childNode);
+            removeElement(node);
+            childNodes.push(node);
+        } else {
+            childNodes.push(childNode);
+        }
+    }
+    target.before(...childNodes);
+}
+
+function removeElement(element) {
+    // manage text as expected in template_inheritance.py
+
+    // const parent = element.parentElement;
+    // const childNodes = [...parent.childNodes];
+    // const index = childNodes.indexOf(element);
+    // const previousText = getTextNode(childNodes.slice(0, index).reverse());
+    // const nextText = getTextNode(childNodes.slice(index));
+    // if (previousText && nextText) {
+    //     previousText.data += nextText.data;
+    //     nextText.remove();
+    // }
+    element.remove();
+}
+
+function getNode(sourceDoc, spec) {
+    if (spec.tagName === "xpath") {
+        // handle errors as expected in template_inheritance.py
+        const result = sourceDoc.evaluate(
+            spec.getAttribute("expr"),
+            sourceDoc.documentElement,
+            null,
+            XPathResult.FIRST_ORDERED_NODE_TYPE
+        );
+        return result.singleNodeValue;
+    }
+
+    if (spec.tagName === "field") {
+        const name = spec.getAttribute("name");
+        return sourceDoc.documentElement.querySelector(`field[name=${name}]`);
+    }
+
+    outer: for (const node of sourceDoc.documentElement.querySelectorAll(spec.tagName)) {
+        for (const { name, value } of spec.attributes) {
+            if (name !== "position" && node.getAttribute(name) !== value) {
+                // not sure about this line
+                continue outer;
+            }
+        }
+        return node;
+    }
+
+    return null;
+}
+
+/**
+ * @param {string} arch
+ * @param {string} inherits
+ * @returns {string}
+ */
+export function applyInheritance(arch, inherits) {
+    const archXmlDoc = parser.parseFromString(arch, "text/xml");
+    const inheritsDoc = parser.parseFromString(inherits, "text/xml");
+
+    const operations = [];
+    for (const child of inheritsDoc.documentElement.children) {
+        if (child.tagName === "data") {
+            for (const c of child.children) {
```

Comment:
in python this exists too but, I don't think it is ideal
have a single xml document that goes:
```xml
<data>
    <xpath .... />
</data>

<xpath ... />
 ```
the xpath instructions in data will execute last, i don't think it is in the [empiric] specs of odoo's xpath resolution, and I do think it may be confusing to have some inheritance at the begining of a document that execute last (to do that, the user should just put his instructions last)